### PR TITLE
[pt1][tensor] New in StaticContext returns at::DataPtr

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#12029 [pt1][tensor] New in StaticContext returns at::DataPtr**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D9889990/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #12105 [pt1][tensor] Split Allocator&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10001033/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #12111 [pt1][tensor] Remove New with Allocator Registry&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10022853/)

In order to remove New() function in StaticContext(to remove StaticContext) and converge to the Allocator design, we'll first change the return type of New to at::DataPtr.

Differential Revision: [D9889990](https://our.internmc.facebook.com/intern/diff/D9889990/)

### CODE REVIEW

File: aten/src/ATen/core/Allocator.h [hunk 1]
```diff
@@ -43,13 +43,19 @@ class DataPtr {
   void* release_context() {
     return ptr_.release_context();
   }
+  std::unique_ptr<void, DeleterFnPtr>&& move_context() {
+    return std::move(ptr_.move_context());
+  }
```
Comment (ezyang) [inline aten/src/ATen/core/Allocator.h]: This is one of those things where I'm still not really sure what idiomatic C++ design here is. I guess this should be OK for now :)

File: caffe2/core/context_gpu.h [hunk 2]
```diff
@@ -334,26 +334,26 @@ inline void CPUContext::CopyBytes<CPUContext, CUDAContext>(
...
   ~PinnedCPUAllocator() override {}
-  std::pair<void*, MemoryDeleter> New(size_t nbytes) override {
+  at::DataPtr allocate(size_t nbytes) const override {
     void* data;
     std::lock_guard<std::mutex> lock(CUDAContext::mutex());
     if (IsNUMAEnabled()) {
-      auto ptr_and_deleter = baseAllocator_.New(nbytes);
-      data = ptr_and_deleter.first;
+      auto data_ptr = baseAllocator_.allocate(nbytes);
+      data = data_ptr.move_context().release();
```
Comment (ezyang) [inline caffe2/core/context_gpu.h]: This looks bad; you need to hold on to the old deleter and then repackage the final pointer with it. Honestly, I would just not `release()` at all; just move the context out, and then move it back in for the reconstruction. (If the problem is that we don't have a `DataPtr` constructor that takes a `unique_ptr` context, we should add one.)

File: caffe2/core/context_gpu_test.cc [hunk 3]
```diff
@@ -11,12 +11,6 @@ CAFFE2_DECLARE_bool(caffe2_cuda_full_device_control);
 
 namespace caffe2 {
 
-namespace {
```
Comment (ezyang) [inline caffe2/core/context_gpu_test.cc]: Good!

File: caffe2/core/hip/context_hip.h [hunk 4]
```diff
@@ -323,26 +323,26 @@ inline void CPUContext::CopyBytes<CPUContext, HIPContext>(
...
   ~PinnedCPUAllocator() override {}
-  std::pair<void*, MemoryDeleter> New(size_t nbytes) override {
+  at::DataPtr allocate(size_t nbytes) const override {
     void* data;
     std::lock_guard<std::mutex> lock(HIPContext::mutex());
     if (IsNUMAEnabled()) {
-      auto ptr_and_deleter = baseAllocator_.New(nbytes);
-      data = ptr_and_deleter.first;
+      auto data_ptr = baseAllocator_.allocate(nbytes);
+      data = data_ptr.move_context().release();
```
Comment (ezyang) [inline caffe2/core/hip/context_hip.h]: Same problem here.

File: caffe2/core/hip/miopen_wrapper.h [hunk 5]
```diff
@@ -26,8 +26,8 @@ struct MIOPENWorkspace
         if(nbytes_ < nbytes)
         {
             reset();
-            auto data_and_deleter = HIPContext::New(nbytes);
-            data_                 = {data_and_deleter.first, data_and_deleter.second};
+            auto data_ptr = HIPContext::New(nbytes);
+            data_ = {data_ptr.move_context().release(), data_ptr.get_deleter()};
```
Comment (ezyang) [inline caffe2/core/hip/miopen_wrapper.h]: If you move out the context, doesn't that also move out the deleter? In that case you're going to get the wrong deleter in the code here.

The root cause of the problem is that you should store a DataPtr. I think that's worth doing in this PR.

File: caffe2/core/qtensor.h [hunk 6]
```diff
@@ -122,10 +122,10 @@ class CAFFE2_EXPORT QTensor {
 
   unsigned char* mutable_data() {
     if (!data_) {
-      auto ptr_and_deleter = Context::New(nbytes());
+      auto data_ptr = Context::New(nbytes());
       data_.reset(
-          static_cast<unsigned char*>(ptr_and_deleter.first),
-          ptr_and_deleter.second);
+          static_cast<unsigned char*>(data_ptr.move_context().release()),
+          data_ptr.get_deleter());
```
Comment (ezyang) [inline caffe2/core/qtensor.h]: Yeah, store a DataPtr here too.

File: caffe2/core/tensor_impl.h [hunk 7]
```diff
@@ -595,29 +618,19 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
...
-        ptr = ptr_and_deleter.first;
-        deleter = ptr_and_deleter.second;
-        storage_.set_data_ptr(at::InefficientStdFunctionContext::makeDataPtr(
-            ptr,
-            [size, dtor, deleter](void* local_ptr) -> void {
-              dtor(local_ptr, size);
-              deleter(local_ptr);
-            },
+        storage_.set_data_ptr(PlacementDeleteContext::makeDataPtr(
+            std::move(data_ptr.move_context()),
```
Comment (ezyang) [inline caffe2/core/tensor_impl.h]: Perhaps, we should make the constructor makeDataPtr accept a DataPtr directly?

File: caffe2/core/context_gpu.h (hunk 2 above)
Comment (jerryzh168) [reply caffe2/core/context_gpu.h]: In this case I'll just do `data = data_ptr.get()` to get the data pointer. let's see if we need a constructor for other places..

File: caffe2/core/hip/miopen_wrapper.h (hunk 5 above)
Comment (jerryzh168) [reply caffe2/core/hip/miopen_wrapper.h]: I was worried to break things in the beginning, but looking at APIs for this one, we should be able to do that.

File: caffe2/core/tensor_impl.h (hunk 7 above)
Comment (jerryzh168) [reply caffe2/core/tensor_impl.h]: sure

File: caffe2/core/context_gpu.h (hunk 2 above)
Comment (ezyang) [reply caffe2/core/context_gpu.h]: Err, just getting the data pointer won't preserve ownership correctly, will it?

File: aten/src/ATen/core/TensorImpl.h [hunk 8]
```diff
@@ -735,29 +767,19 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
...
       } else {
         // For fundamental type, new and delete is easier.
-        auto ptr_and_deleter =
-            GetStaticContext()->New(numel_ * storage_.itemsize());
-        storage_.set_data_ptr(at::InefficientStdFunctionContext::makeDataPtr(
-            ptr_and_deleter.first,
-            ptr_and_deleter.second,
-            at::Device(storage_.device_type())));
+        auto data_ptr = GetStaticContext()->New(numel_ * storage_.itemsize());
+        storage_.set_data_ptr(std::move(data_ptr));
```
Comment (ezyang) [inline aten/src/ATen/core/TensorImpl.h]: nit: probably just better to inline `data_ptr` away here :)

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**